### DISCUSSION TITLE
could you use lf as a filechooser, like ranger with the --choosefile argument?

### DISCUSSION THREAD

**Discussion by emilBeBri** [gokcehan/lf#1080] [Ideas]
So I'd like to use lf as my filechooser -  especially for  qutebrowser.   

Perhaps it's already possible to make lf output the files selected in some way? basically you just need to spawn lf and let the A output when you quit be the files selected.  but the best would probably be the something like

--choosefile 

and 

--choosefiles

from ranger, so that in the first mode, only one file can be selected, and in the second, multiple can be selected.

If someone tells me how to, I'd be happy to add a 'Qutebrowser' section in the wiki with the shell script for this, and also how to put it qutebrowsers config file.

**Comment by emilBeBri** [comment]
anyone?

**Comment by Joyje** [comment]
You should read the instructions for lf. The selected files are stored in 3 different variables. $f for the currently marked file (single file), $fs for the current selection of files, and $fx for choosing $f if there are no selections and $fs if there are. You can read about it [here](https://pkg.go.dev/github.com/gokcehan/lf#section-readme).

**Reply by emilBeBri** [reply]
I have read the instructions of course.  And I use those vars for lots of different stuff But I still can't wrap my head around how I would use it to select a files in a filepicker.

**Reply by Joyje** [reply]
Maybe I misunderstood your question. I don't know what qutebrowser is and I'm not sure as to what you mean by filechooser. Referring to ranger isn't helpful either since I never used ranger and don't know anything about what it can or can't do. I use lf a whole lot though, so maybe if you specify in more detail what you mean by "filechooser" I could help you out, unless the answer provided by @gokcehan already solved it for you.

**Reply by emilBeBri** [reply]
Ok, maybe I was a bit unspecific - I think gokchen have answered it, but for some reason it's not working even though it should be fairly straightforward, so I think he can probably clarify it.

What I'm looking for is way to use lf to select files in my webbrowser - in this example, qutebrowser ,  but it's the same in other browsers: there is a button that says 'upload files from your computer', or something like that, like gmail attachments, and I'd like to use lf to choose the files. I made a bash script that does this with fzf, but most times I find it a lot more easy to navigate to the files - right now I'm forced to using pcman, which is ok, but nowhere near lf.

**Reply by Joyje** [reply]
Huh! Sounds a bit tricky. I never used `lf` in that way, but I understand you would want to send the file paths for all selected files to your browsers input somehow.

I don't know how your browser reads file paths, or how it gets the path data, but in order to export the file paths to all selected files from `lf` you can do this in  your `lfrc`:

`cmd export-path ${{`
`    echo $fx >> /file/with/exported/paths`
`}}`

If you want all of the file paths in one string separated by your `IFS` (single input), or:

`cmd export-path ${{`
`    for file in $fx; do`
`        echo $file >> /file/with/exported/paths`
`    done`
`}}`

If you want every file path on a new line (multiple inputs, one per path).

Don't forget to:

`map <key> export-path`

If your browser expects a string with all the file paths, make sure what you specified in your `IFS` is compatible with how your browser reads that string. You can change the separator `lf` uses with `set filesep "Input_desired_sep_here"`.

After testing to see that it works and if the format is correct, you would of course change `/file/with/exported/paths` to the process expecting the input to your browser. You could then also change your mapping to `map <key> :export-path; quit` so `lf` closes after choosing the files.

**Comment by gokcehan** [comment]
@emilBeBri You are probably looking for `-selection-path` option. You can see the full list of command line options with `lf -help`.

**Reply by emilBeBri** [reply]
Hi ok thanks, tried the docs [here](https://pkg.go.dev/github.com/gokcehan/lf#hdr-Quick_Reference) but it's not mentioned. 


So I'm a bit confused to select those files - I use space to select files/dirs, and can den yank, paste, delete, and run shell scripts on them. But when I try:

```

lf -selection-path /home/emil/test_lf

```
none of the selected files gets written to that file. 

As per [this issue in the tracker](https://github.com/gokcehan/lf/issues/89), you should be able to write to a bash variable like this:


```
filename=$(lf -selection-path /dev/stdout)

```
however nothing is in there when I echo it. 

There must be something I'm doing wrong here. 

lf -version says '28'. On manjaro linux


I'd be happy to put this in the wiki when the solution is clear.

**Reply by veltza** [reply]
After selecting files, you need to open them (default 'l' and 'right' keys).

**Reply by Joyje** [reply]
@emilBeBri I tried this hastily and didn't get it to work either at first. After trying it again just now, properly this time, I got it working. It seems like this works more or less the exact same way as the code I provided in my previous comment above. The difference being that with `-selection-path` the function is directly replacing your open command, so you would use "i", "l" or "right" by default to run it, not needing to map the function to a new key. Thanks to @veltza for pointing it out.

This answer should be the correct one for your question. However, if your browser require the paths to be written in a specific way that isn't the default format provided by `-selection-path`, my comment above gives some insight as to how one might manually create the desired format. Let us know if you got it to work.

**Reply by emilBeBri** [reply]
@Joyje  and co, great now I got it, thank you. After some trial and error I got it working by making a script like this:



```
#!/bin/bash
# qb-fileselect-lf.sh
# set euo pipefail

case "$1" in
	file)
		thetype="file"
		;;
	files)
		thetype="file"
		;;
esac


tmpfile=$(mktemp)
CMD="lf -selection-path $tmpfile"
kitty  sh -c "$CMD" 



line_count=$( grep -c '^' $tmpfile)
# lolecho $line_count
# Test if the file has more than one line
if [ $line_count -gt 1 ] && [ thetype == 'file' ]; then
    notify-send -u critical -i "you choose > 1 file and you should only choose one"
    exit 0
fi

cat $tmpfile


```

which can be fed into the qutebrowser config python script like this - I know that probably won’t mean much to you here but for completeness sake I'll post it.   

```
c.fileselect.multiple_files.command = ['qb-fileselect-lf.sh', 'files']
c.fileselect.single_file.command = ['qb-fileselect-lf.sh', 'file']
c.fileselect.folder.command = ['notify-send', '"folder command. not working ATM. fix it somehow when you get to it." ']

```

There are three ways of handling opening files within qutebrowser , 'multiple files',  'single file' or 'folder'. 

It's working with file and files. At the moment  I just put in a notify-send when a folder-opening event is triggered, and I'll deal with it then, easier to handle when you have a a concrete case. 

Now I'm not sure how to handle webpages that want me to upload a folder, since as it is by using 'l' or right arrow you go into that folder, so this might complicate things. But perhaps I can just map something like <Leader>d,  so that when it's a folder, I can send the path of the folder to sdout or something. not quite sure how to do that but I'll figure it out - If you can show me an elegant way of doing it - or a better way - that would be nice. 

 I'll make a section in the wiki when that functionality is ready. Thank you for your help!

**Comment by sudoforge** [comment]
i see that you came to a solution; as a fellow `qutebrowser` + `lf` user, i thought i'd just leave my configuration here for you. i don't do anything fancy, and yes, folder selection is broken (but i haven't really ever attempted to upload an entire folder, usually just uploading individual files or archive files).

```
# Configure the filepicker
filepicker = [
    "alacritty",
    "--class",
    "filepicker",
    "--title",
    "filepicker",
    "-e",
    "lf",
    "-command",
    "set nohidden",
    "-selection-path={}",
]
c.fileselect.handler = "external"
c.fileselect.folder.command = filepicker
c.fileselect.multiple_files.command = filepicker
c.fileselect.single_file.command = filepicker
```

_[source](https://github.com/sudoforge/dotfiles/blob/899a2db60ae35dca5b42f5e9733497eb589cf511/qutebrowser/.config/qutebrowser/config.py#L170-L186)_

i have filepicker actions open up a new terminal window that is sized to my preferences through window and title classes (using bspwm):

```
bspc rule -a filepicker:filepicker:\* state=floating rectangle=1200x800+510+352
```

i think i'll steal your `notify-send` for `c.fileselect.folder.command`.

**Reply by emilBeBri** [reply]
Nice looks good! Nice with someonehaving the same very particular taste in softwareas I do - haven't had a reason to get the folder picker working yet,  if I do I'll post it here and if you do first that would be greatly appreciated! 

Is there a reason you've chosen bspwm as wm instead of i3? I have only tried that and just stuck with it but wondering if I sound move on at some point

**Reply by sudoforge** [reply]
> Is there a reason you've chosen bspwm as wm instead of i3?

i chose `bspwm` over `i3` strictly because it is lightweight, single-purpose, and doesn't add additional bulk that i already had other solutions for (e.g. keybinding, a bar). that said, i'm moving from x to wayland in the next few months, which means revisiting the WM i use. i may end up with `sway` (i3-ish), `river`, or some other compositor.

**Reply by sudoforge** [reply]
i mean, it's also "more scriptable" and "faster" (according to people who have used both).
